README.md
# Introduction to OpenGL

- Building
  - [Windows](#windows)
  - [Linux](#linux)
  - [MacOs](#macos)

---

## Windows

### Prerequisites

Not necessary if you already did the previous tp, but to recall:

- download and install the latest version of CMake

  - download here (choose "Windows x64 Installer:"): <https://cmake.org/download/>

> [!IMPORTANT]
> When installing make sure that the checkbox "ne pas ajouter cmake au PATH" is NOT checked

- if you don't have it already, download and install MS Visual Studio Community Edition (free for students): <https://visualstudio.microsoft.com/downloads/>

  - install instructions here: <https://docs.microsoft.com/en-us/cpp/build/vscpp-step-0-installation?view=vs-2019>

> [!IMPORTANT]
> install the "Desktop development with C++"
  
  - If you have VS already installed, you can go in **Tools** --> **Get Tools and Features...** to install "Desktop development with C++" if it is missing.

### Create the Visual Studio Solution

This step enables you to create the project file to load inside VS:

- unzip the code inside a folder. *Avoid to place the code in folders with spaces and accented characters*.

- open a Terminal and go to the directory containing the code.

- execute:

  - `md build`
  
  - `cd build`
  
  - `cmake -G "Visual Studio 16 2019" -A x64 -DCMAKE_BUILD_TYPE:STRING=Release ..`
  
  - `dir`
  
> [!NOTE]
> if you had a different version of VS installed (not the latest) you may need to adapt the string `Visual Studio 16 2019` to your version: e.g. Visual Studio 15 2017, Visual Studio 14 2015, Visual Studio 12 2013
  
- if everything went well you should find a file named `tp2.sln` inside the directory.

### Compile, build, execute

- open `tp1.sln` inside VS either by double clicking on it or opening from inside VS

- build the solution (**Build Solution** from the **Build menu**)

- from the tp directory copy `freeglut\bin\x64\freeglut.dll` in `build\Release`

- execute the code:  

  - Select the project you want to run (e.g. `robot`), right click on it and select **Set as Startup Project**
  
  - On the menu bar, choose **Debug** --> **Start without debugging**.)

(see <https://docs.microsoft.com/en-us/cpp/build/vscpp-step-2-build?view=vs-2019> for how to build, execute, etc)

### Editing the code

Edit the code according to the assignments that are given, rebuild the solution and execute.

> [!NOTE]
> You need to run the cmake line only **once**

> [!NOTE]
> You need to copy the dll file only **once**.

---

## Linux

### Prerequisites

**Not necessary if you already did the previous tp.**

In order to develop with OpenGL some system packages are required (unless you are using the N7 machines):

```shell
sudo apt-get install libglu1-mesa-dev freeglut3-dev build-essential mesa-common-dev libxi-dev libxmu-dev automake
```

To build this code we use the CMake build system. You can install CMake from the system package manager but you need a recent version >= 3.10. Check the version that is provided by your linux distribution and if it is suitable usually you need to

```shell
sudo apt-get install cmake
```

otherwise you can install the binaries from here (choose Linux x86_64): <https://cmake.org/download/>

Once downloaded, in order to install:

```shell
chmod +x cmake-X.YY.Z-Linux-x86_64.sh
sudo ./cmake-X.YY.Z-Linux-x86_64.sh --prefix=/usr/local/ --skip-license
```
  
### Build

To compile and build the code you do 

 ```shell
 mkdir build && cd build
 cmake ..
 make robot
 ```

Also,

```shell
make all
```

builds everything, and

```shell
make clean
```

cleans everything.

Execute the code:

```shell
./robot
```

### Editing the code

Edit the code as required and then

```shell
make robot
```

> [!NOTE]
> the cmake line has to be run only **once**

---

## macOS

### Prerequisites

In order to develop with OpenGL check if

```shell
ls  /System/Library/Frameworks/
```

contains OpenGL and GLUT frameworks.
If not you need to install XCode  from the `Mac App Store`, see here for more details <https://developer.apple.com/support/xcode/>

Download and install the latest version of CMake <https://cmake.org/download/> OR if you have `brew` installed you can do

```shell
brew install cmake
```

### Build

Same as Linux.

### Editing the code

Same as Linux.


changelog.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [2023.1.0]

### Added

- warning flags for gcc, clang and msvc
- use of studentipy tags in the code
- script to make release with studentipy
- CI on GitHub for linux, windows and osx
- clang format file
- added IPO support 

### Changed

- modernize and refactoring cmake
- switched to c++17

### Fixed

- all warnings on linux and windows

### Removed

- support for makefile building
- appveyor ci support

## [2021.0.0] - 2021-05-05

### Fixed

- comments in the code
- cmakefile using targets for glut

## [2020.1.0] - 2020-05-05

First release on GitHub

### Added

- cross-platform support
- freeglut for windows
- appveyor ci support


robot.c

#include <stdlib.h>
#include <stdio.h>

// for mac osx
#ifdef __APPLE__
#include <OpenGL/gl.h>
#include <OpenGL/glu.h>
#include <GLUT/glut.h>
#else
// only for windows
#ifdef _WIN32
#include <windows.h>
#endif
// for windows and linux
#include <GL/gl.h>
#include <GL/glu.h>
#include <GL/freeglut.h>
#endif

#include <math.h>

// Global variables to animate the robotic arm
int Angle1 = 0;
int Angle2 = 0;

// Global variables to rotate the arm as a whole
int RobotAngleX = 0;
int RobotAngleY = 0;



//*************************************************************************
// Function that draws a reference system
//*************************************************************************
void DrawReferenceSystem()
{
    //**********************************
    // set the line width to 3.0
    //**********************************

    //**********************************
    // Draw three lines along the x, y, z axis to represent the reference system
    // Use red for the x-axis, green for the y-axis and blue for the z-axis
    //**********************************




    //**********************************
    // reset the drawing color to white
    //**********************************

    //**********************************
    // reset the line width to 1.0
    //**********************************

}


//*************************************************************************
// Function that draws a single joint of the robotic arm
//*************************************************************************
void DrawJoint()
{
    // first draw the reference system
    DrawReferenceSystem();

    // Draw the joint as a parallelepiped (a cube scaled on the y-axis)
    // the bottom face of the cube must be on the xz plane



    //**********************************
    // draw the parallelepiped
    //**********************************




}

// Function that draws the robot as three parallelepipeds
void DrawRobot()
{
    //**********************************
    // It's better to work on a local reference system...
    //**********************************

    // draw the first joint
    DrawJoint();

    // Draw the other joints
    // every joint must be placed on top of the previous one
    // and rotated according to the relevant Angle
    //**********************************
    // the second joint
    //**********************************



    //**********************************
    // the third joint
    //**********************************



    //**********************************
    // "Release" the copy of the current MODELVIEW matrix
    //**********************************

}


//*************************************************************************
// display callback
//*************************************************************************
void display(void)
{
    // clear the window
    glClear (GL_COLOR_BUFFER_BIT);

    // working with the GL_MODELVIEW Matrix
    glMatrixMode(GL_MODELVIEW);

    //**********************************
    // we work on a copy of the current MODELVIEW matrix, hence we need to...
    //**********************************


    //**********************************
    // Rotate the robot around the x-axis and y-axis according to the relevant angles
    //**********************************



    // draw the robot
    DrawRobot();

    //**********************************
    // not anymore on a local reference system
    //**********************************

    // flush drawing routines to the window
    glutSwapBuffers();
}

//*************************************************************************
// Special keys callback
//*************************************************************************
void arrows (int key, int x, int y)
{
        //**********************************
        // Manage the update of RobotAngleX and RobotAngleY with the arrow keys
        //**********************************




        glutPostRedisplay ();
}

//*************************************************************************
// Keyboard callback
//*************************************************************************
void keyboard (unsigned char key, int x, int y)
{
    switch (key)
    {
        case 'q':
        case 27:
            exit(0);
        break;
        //**********************************
        // Manage the update of Angle1 with the key 'a' and 'z'
        //**********************************







        //**********************************
        // Manage the update of Angle2 with the key 'e' and 'r'
        //**********************************








        default:
        break;
    }
    glutPostRedisplay ();
}



void init(void)
{
    glClearColor (0.0, 0.0, 0.0, 0.0);
    glMatrixMode (GL_PROJECTION);
    glLoadIdentity ();
    gluPerspective(65.0, 1.0, 1.0, 100.0);

    glShadeModel (GL_FLAT);
    glMatrixMode(GL_MODELVIEW);
    glLoadIdentity();

    // Place the camera
    gluLookAt(-6,5,-6,0,0,2,0,1,0);
}


//*************************************************************************
// Function called every time the main window is resized
//*************************************************************************
void reshape ( int width, int height )
{

    // define the viewport transformation;
    glViewport(0,0,width,height);
    if (width < height)
        glViewport(0,(height-width)/2,width,width);
    else
        glViewport((width-height)/2,0,height,height);
}


//*************************************************************************
// Prints out how to use the keyboard
//*************************************************************************
void usage()
{
    printf ("\n*******\n");
    printf ("Arrows key: rotate the whole robot\n");
    printf ("[a][z] : move the second joint of the arm\n");
    printf ("[e][r] : move the third joint of the arm\n");
    printf ("[esc]  : terminate\n");
    printf ("*******\n");
}

//////////////////////////////////////////////////////////////
int main(int argc, char** argv)
{
   glutInit(&argc, argv);
   glutInitDisplayMode (GLUT_DOUBLE | GLUT_RGB);
   glutInitWindowSize (500, 500);
   glutInitWindowPosition (100, 100);
   glutCreateWindow(argv[0]);
   init ();
   glutDisplayFunc(display);

   glutReshapeFunc(reshape);
    //**********************************
   // Register the keyboard function
   //**********************************

   //**********************************
   // Register the special key function
   //**********************************

   // just print the help
   usage();

   glutMainLoop();

   return EXIT_SUCCESS;
}

//////////////////////////////////////////////////////////////


generateStudent.py
#!/usr/bin/env python3

import argparse
import logging
import os
import shutil
import subprocess
import sys
import tempfile
import zipfile

logging.basicConfig(
    level=logging.INFO,
    format="[%(levelname)s] %(message)s"
)
logger = logging.getLogger()


def main(working_dir: str, archive_name: str, skip_cleaning: bool = False, skip_packaging: bool = False):

    if not os.path.exists(working_dir):
        logger.info("Creating working directory %s", working_dir)
        os.makedirs(working_dir)

    dest_name = os.path.join(working_dir, archive_name)

    logger.info("Creating directory %s", dest_name)
    os.mkdir(dest_name)

    files_to_copy = ["CMakeLists.txt", "README.md", "LICENSE", "robot.cpp", ".clang-format"]
    logger.info("Copying files to %s", dest_name)
    for file in files_to_copy:
        shutil.copy(file, dest_name)

    logger.info("Copying directories data and freeglut to %s", dest_name)
    shutil.copytree("freeglut", os.path.join(dest_name, "freeglut"))

    studentify_dir = os.path.join(tempfile.gettempdir(), "tpt")
    logger.info("Cloning studentify.py to %s", studentify_dir)
    subprocess.run(["git", "clone", "https://github.com/simogasp/studentipy.git", studentify_dir])

    robot_cpp = os.path.join(dest_name, "robot.cpp")
    logger.info("Applying studentify to %s", robot_cpp)
    subprocess.run(
        ["python3", os.path.join(studentify_dir, "studentify.py"), robot_cpp, "-o", robot_cpp, "--force"])

    logger.info("Removing studentify.py")
    shutil.rmtree(studentify_dir)

    if not skip_packaging:
        logger.info("Generating archive %s.zip in %s", archive_name, working_dir)
        with zipfile.ZipFile(os.path.join(working_dir, f"{archive_name}.zip"), "w") as zip_file:
            for root, _, files in os.walk(dest_name):
                for file in files:
                    zip_file.write(os.path.join(root, file), os.path.relpath(os.path.join(root, file), dest_name))

    if skip_cleaning:
        return
    logger.info("Cleaning up %s", dest_name)
    shutil.rmtree(dest_name)


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Script used to generate the package to be given to the student.")
    parser.add_argument(
        "-v",
        "--verbosity",
        dest="verbosity",
        choices=["DEBUG", "INFO", "WARNING", "ERROR", "CRITICAL"],
        default="INFO",
        help="Set the logging verbosity level (choose from DEBUG, INFO, WARNING, ERROR, CRITICAL).",
    )
    parser.add_argument(
        "-w",
        "--working-dir",
        dest="working_dir",
        help="Set the working directory where the files will be created.",
    )
    parser.add_argument(
        "-a",
        "--archive-name",
        dest="archive_name",
        help="The name of the archive to be created. [default=tp2ogl-v<version>]"
    )
    parser.add_argument(
        "--skip-cleaning",
        action = 'store_true',
        dest="skip_cleaning",
        help="Skip the cleaning step leaving the code available.",
    )
    parser.add_argument(
        "--skip-packaging",
        action = 'store_true',
        dest="skip_packaging",
        help="Skip the cleaning step leaving the code available.",
    )

    args = parser.parse_args()

    if args.verbosity:
        logger.setLevel(args.verbosity)

    if args.working_dir:
        wdir = args.working_dir
    else:
        wdir = tempfile.gettempdir()

    if args.archive_name:
        zip_name = args.archive_name
    else:
        version = subprocess.run(["git", "describe", "--tags"], check=True, stdout=subprocess.PIPE).stdout.decode("utf-8").strip()[1:]
        zip_name = f"tp2ogl-v{version}"

    try:
        main(working_dir=wdir, archive_name=zip_name, skip_cleaning=args.skip_cleaning, skip_packaging=args.skip_packaging)
    except Exception as e:
        logger.exception(e)
        sys.exit(1)


robot.cpp

#include <cstdlib>
#include <cstdio>
#include <cmath>
#include <algorithm>

// for mac osx
#ifdef __APPLE__
#include <OpenGL/gl.h>
#include <OpenGL/glu.h>
#include <GLUT/glut.h>
#else
// only for windows
#ifdef _WIN32
#include <windows.h>
#endif
// for windows and linux
#include <GL/gl.h>
#include <GL/glu.h>
#include <GL/freeglut.h>

#endif


// Global variables to animate the robotic arm
float angle1 = 0;
float angle2 = 0;

// Global variables to rotate the arm as a whole
float robotAngleX = 0;
float robotAngleY = 0;

//<!!
// how open are the pincers [0,1]
constexpr float START{0.125f};
constexpr float END{0.5f};
float open = 1.0f;
//>!!

/**
 * Function that draws the reference system (three lines along the x, y, z axis)
 */
void drawReferenceSystem()
{
    //**********************************
    // set the line width to 3.0
    //**********************************
    glLineWidth(3.f);   //!!
    //**********************************
    // Draw three lines along the x, y, z axis to represent the reference system
    // Use red for the x-axis, green for the y-axis and blue for the z-axis
    //**********************************
//<!!
    glBegin(GL_LINES);
    //**********************************
    // Set the color red for the x-axis
    //**********************************
    glColor3f(1.f, 0.f, 0.f);
    //**********************************
    // Set TWO ending points for the line representing the x-axis
    //**********************************
    glVertex3f(0.f, 0.f, 0.f);
    glVertex3f(1.f, 0.f, 0.f);
    //**********************************
    // Set the color green for the y-axis
    //**********************************
    glColor3f(0.f, 1.f, 0.f);
    //**********************************
    // Set TWO ending points for the line representing the y-axis
    //**********************************
    glVertex3f(0.f, 0.f, 0.f);
    glVertex3f(0.f, 1.f, 0.f);
    //**********************************
    // Set the color blue for the z-axis
    //**********************************
    glColor3f(0.f, 0.f, 1.f);
    //**********************************
    // Set TWO ending points for the line representing the z-axis
    //**********************************
    glVertex3f(0.f, 0.f, 0.f);
    glVertex3f(0.f, 0.f, 1.f);
    //**********************************
    // End the drawing
    //**********************************
    glEnd();
//>!!
    //**********************************
    // reset the drawing color to white
    //**********************************
    glColor3f(1.f, 1.f, 1.f);  //!!
    //**********************************
    // reset the line width to 1.0
    //**********************************
    glLineWidth(1.f);  //!!
}


/**
 * Function that draws a single joint of the robotic arm
 */
void drawJoint()
{
    // first draw the reference system
    drawReferenceSystem();

    // Draw the joint as a parallelepiped (a cube scaled on the y-axis)
    //**********************************
    // Bring the cube "up" so that the bottom face is on the xz plane
    //**********************************
    glTranslatef(0.f, 1.f, 0.f); //!!

    //**********************************
    // draw the scaled cube. Remember that the scaling has to be only
    // on the local reference system, hence we need to get a local copy
    // of the modelview matrix...
    //**********************************
//<!!
    glPushMatrix();
    glScalef(1.f, 2.f, 1.f);
    glutWireCube(1.f);
    glPopMatrix();
//>!!
}

/**
 * Function that draws the robot as three parallelepipeds
 */
void drawRobot()
{
    //**********************************
    // we work on a copy of the current MODELVIEW matrix, hence we need to...
    //**********************************
    glPushMatrix();  //!!

    // draw the first joint
    drawJoint();
    glTranslatef(0.f, 1.f, 0.f);  //!!

    // Draw the other joints: every joint must be placed on top of the previous one
    // and rotated according to the relevant Angle
    //**********************************
    // the second joint
    //**********************************
    glRotatef(angle1, 1.f, 0.f, 0.f);   //!!
    drawJoint();                        //!!
    //**********************************
    // the third joint
    //**********************************
//<!!
    glTranslatef(0.f, 1.f, 0.f);
    glRotatef(angle2, 1.f, 0.f, 0.f);
    drawJoint();

    // Draw the Pincers
    // First the base
    glTranslatef(0.f, 1.125f, 0.f);
    glPushMatrix();
    glScalef(1.f, 0.25f, 0.25f);
    glutWireCube(1);
    glPopMatrix();
    glTranslatef(0.f, 0.45f, 0.f);
    // The the fork
    glPushMatrix();
    glTranslatef(-START + open * (START - END), 0.f, 0.f);
    glScalef(0.25f, 0.65f, 0.25f);
    glutWireCube(1);
    glPopMatrix();
    glPushMatrix();
    glTranslatef(START + open * (END - START), 0.0f, 0.f);
    glScalef(0.25f, 0.65f, 0.25f);
    glutWireCube(1);
    glPopMatrix();
//>!!
    //**********************************
    // "Release" the copy of the current MODELVIEW matrix
    //**********************************
    glPopMatrix();  //!!

}


/**
 * Function that handles the display callback (drawing routine)
 */
void display()
{
    // clear the window
    glClear(GL_COLOR_BUFFER_BIT);

    // working with the GL_MODELVIEW Matrix
    glMatrixMode(GL_MODELVIEW);

    //**********************************
    // we work on a copy of the current MODELVIEW matrix, hence we need to...
    //**********************************
    glPushMatrix();           //!!
    drawReferenceSystem();    //!!
    //**********************************
    // Rotate the robot around the x-axis and y-axis according to the relevant angles
    //**********************************
    glRotatef(robotAngleX, 1.f, 0, 0.f);    //!!
    glRotatef(robotAngleY, 0.f, 1.f, 0.f);  //!!

    // draw the robot
    drawRobot();

    //**********************************
    // "Release" the copy of the current MODELVIEW matrix
    //**********************************
    glPopMatrix();  //!!

    // flush drawing routines to the window
    glutSwapBuffers();
}


/**
 * Function that handles the special keys callback
 * @param[in] key the key that has been pressed
 * @param[in] x the mouse in window relative x-coordinate when the key was pressed
 * @param[in] y the mouse in window relative y-coordinate when the key was pressed
 */
void arrows(int key, int, int)
{
    //**********************************
    // Manage the update of RobotAngleX and RobotAngleY with the arrow keys
    //**********************************
//<!!
    switch (key) {
        case GLUT_KEY_UP:
            robotAngleX = std::fmod(robotAngleX + 5, 360.f);
            break;
        case GLUT_KEY_DOWN:
            robotAngleX = std::fmod(robotAngleX - 5, 360.f);
            break;
        case GLUT_KEY_LEFT:
            robotAngleY = std::fmod(robotAngleY + 5, 360.f);
            break;
        case GLUT_KEY_RIGHT:
            robotAngleY = std::fmod(robotAngleY - 5, 360.f);
            break;
        default:
            break;
    }
//>!!
    glutPostRedisplay();
}


/**
 * Function that handles the keyboard callback
 * @param key  the key that has been pressed
 * @param[in] x the mouse in window relative x-coordinate when the key was pressed
 * @param[in] y the mouse in window relative y-coordinate when the key was pressed
 */
void keyboard(unsigned char key, int, int)
{
    switch (key) {
        case 'q':
        case 27:
            exit(0);
            break;
        //**********************************
        // Manage the update of Angle1 with the key 'a' and 'z'
        //**********************************
//<!!
        case 'a':
            angle1 = std::fmod(angle1 + 5, 360.f);
            // printf("Angle 1: %f\n", angle1);
            break;
        case 'z':
            angle1 = std::fmod(angle1 - 5, 360.f);
            // printf("Angle 1: %f\n", angle1);
            break;
        //**********************************
        // Manage the update of Angle2 with the key 'e' and 'r'
        //**********************************
        case 'e':
            angle2 = std::fmod(angle2 + 5, 360.f);
            // printf("Angle 2: %f\n", angle2);
            break;
        case 'r':
            angle2 = std::fmod(angle2 - 5, 360.f);
            // printf("Angle 2: %f\n", angle2);
            break;

        //**********************************
        // Manage the pincers
        //**********************************
        case 'o':
            open += 0.05f;
            break;
        case 'l':
            open -= 0.05f;
            break;
//>!!
        default:
            break;
    }
    open = std::clamp(open, 0.f, 1.f);   //!!
    glutPostRedisplay();
}


void init()
{
    glClearColor(0.f, 0.f, 0.f, 0.f);
    glMatrixMode(GL_PROJECTION);
    glLoadIdentity();
    gluPerspective(65.0, 1.0, 1.0, 100.0);

    glShadeModel(GL_FLAT);
    glMatrixMode(GL_MODELVIEW);
    glLoadIdentity();

    // Place the camera
    gluLookAt(-1., 5., -1., .0, .0, 2., .0, 1., .0);
}


/**
 * Function called every time the main window is resized
 * @param[in] width the new window width in pixels
 * @param[in] height the new window height in pixels
 */
void reshape(int width, int height)
{

    // define the viewport transformation;
    glViewport(0, 0, width, height);
    if (width < height)
        glViewport(0, (height - width) / 2, width, width);
    else
        glViewport((width - height) / 2, 0, height, height);
}


/**
 * Function that prints out how to use the keyboard
 */
void usage()
{
    printf("\n*******\n");
    printf("Arrows key: rotate the whole robot\n");
    printf("[a][z] : move the second joint of the arm\n");
    printf("[e][r] : move the third joint of the arm\n");
    printf("[o][l] : move the pincers\n"); //!!
    printf("[esc]  : terminate\n");
    printf("*******\n");
}


int main(int argc, char **argv) {
    glutInit(&argc, argv);
    glutInitDisplayMode(GLUT_DOUBLE | GLUT_RGB);
    glutInitWindowSize(500, 500);
    glutInitWindowPosition(100, 100);
    glutCreateWindow(argv[0]);
    init();
    glutDisplayFunc(display);

    glutReshapeFunc(reshape);
    //**********************************
    // Register the keyboard function
    //**********************************
    glutKeyboardFunc(keyboard);  //!!

    //**********************************
    // Register the special key function
    //**********************************
    glutSpecialFunc(arrows);  //!!

    // just print the help
    usage();

    glutMainLoop();

    return EXIT_SUCCESS;
}


